src/inmanta/data/__init__.py

```diff
@@ -4483,6 +4483,33 @@ async def trim(cls, environment: UUID, connection: Optional[Connection] = None)
             connection=connection,
         )
 
+    @classmethod
+    async def get_aggregated_deployment_status(
+        cls, environment_id: UUID, resource_ids: list[ResourceIdStr]
+    ) -> dict[NonDeployingResourceState, int]:
+        """
+        Fetches aggregated deployment status counts for a list of resource IDs within a given environment.
+
+        :param environment_id: UUID of the environment.
+        """
+        query = f"""
+        SELECT last_non_deploying_status, COUNT(*) AS status_count
+        FROM {cls.table_name()}
+        WHERE environment = $1 AND resource_id = ANY($2)
+        GROUP BY last_non_deploying_status
+        """
+        values = [environment_id, resource_ids]
+
+        status_counts = {}
+        async with cls.get_connection() as con:
+            async with con.transaction():
```

Comment:
for this one, it may be more interesting to use [fetch](https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.fetch) instead of a cursor? (there will never be many records in the result)

Reply:
It used to be like this:
```
        query = f"""
        SELECT last_non_deploying_status, COUNT(*) AS status_count
        FROM {cls.table_name()}
        WHERE environment = $1 AND resource_id = ANY($2)
        GROUP BY last_non_deploying_status
        """
        results = await cls._fetch_query(query, environment_id, resource_ids)

        status_counts = {
            cast(str, record["last_non_deploying_status"]): cast(int, record["status_count"]) for record in results
        }
        return status_counts
```
 @arnaudsjs thought a cursor would be better. I have no idea how many records there will be and what we consider to be many.

Reply:
The default prefetch size of a cursor is 50 elements. So as far as I can see we will fetch everything in a single batch if the number of records is low. But I have no strong objections against using fetch(), so if Wouter prefers that one, it's fine for me as well.

Reply:
The method is now different, so I don't think the fetch is still a benefit (it used to return < 10 resources, now it is a lot more)